rfcs/20220201-quantization-scaling/README.md

```diff
@@ -0,0 +1,388 @@
+        const_dnnl_post_ops_t post_ops, int index, float *scale,
+        dnnl_alg_kind_t *alg_kind, const dnnl_memory_desc_t **src1_desc);
+```
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(scale, algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+```
+
+#### 2.2.2 Option 2
+
+An alternative solution would be to extend mechanism `attr::scales`. This
+solution doesn't require new API, but extends the existing one.
+
+An example:
+
+```cpp
+    post_ops p_ops;
+    p_ops.append_binary(algorithm::binary_add, summand_md);
+
+    primitive_attr attr;
+    attr.set_post_ops(p_ops);
+    attr.set_scales(DNNL_ARG_ATTR_MULTIPLE_POST_OP(0) | DNNL_ARG_SRC_1,
+            /* mask */ 0, {DNNL_RUNTIME_F32_VAL});
+```
+
+While the alternative solution allows to avoid new API, it introduces one more
+way to set scale which complicates both implementation and integration.
+Please review section
+[Unifying scaling in long-term](#3-unifying-scaling-in-long-term) for more
+details.
+
+
+### 2.3. Scales for inputs in Concat
+
+The mechanism `attr:set_scales` will be introduced to Concat primitive. This
+mechanism is used in Binary primitive, so API is already there.
+
+The following limitations are in place for an initial implementation:
+ - no-runtime support;
+ - mask is 0.
+
+An example:
+
+```cpp
+    primitive_attr attr;
+    for (int n = 0; n < num_src; ++n)
+        attr.set_scales(DNNL_ARG_MULTIPLE_SRC + n, /* mask */ 0, src_scales[n]);
+```
+
+## 3. Unifying scaling in long-term (oneDNN v3.0)
+
+The motivation behind unifying scaling:
+- Remove duplications;
+- Provide an aligned support across all use cases.
+
+There are a few issues with the current scaling support:
+ - Output scale, scales and binary_mul overlap each other;
+ - RNN quantization is specific to RNN, but can be replaced by other
+ scaling mechanisms:
+   ```
+   # Scaling semantics of set_scales and set_zero_points
+   x_f32 = scale * (x_i8 - zero_point)
+
+   # Scaling semantics of RNN quantization
+   x_i8 = scale * x_f32 + zero_point
+   x_f32 = 1 / scale * (x_i8 - zero_point)
+   ```
+   This will affect application side code as user should pass `1 / scale` to
+   oneDNN comparing to the current `scale`;
+ - Built-in parameter provides very limited flexibility (no mask, no runtime).
+
+### The changes:
+
+#### New quantization scheme
+
+Formula:
+```
+dst = post_ops(OP(...)/scale_dst + zp_dst)
+```
+
+New quantization improves integration by removing overhead of:
+- pre-computing output scale
+- bias quantization
+
+The following images show differences between existing quantization and new quantization targeting v3
+
+![Quantization in oneDNN v2](quantization-v2.png)
+
+![Quantization in oneDNN v3](quantization-v3.png)
+
+Notes:
+- scales/zero_points are applied to their respective argument
+- if a users needs re-scaling/re-shifting in the post chain, they can rely on
+  binary post-op
```

Comment:
This is contradicts what is below, where scale/zero_point of post-ops are passed through set_scales/set_zero_points. Using the attribute mecanism seems to be the best option as we discussed, since it allows to reduce the number of post-ops in post-op chain, and allows to reuse existing code to apply bias without adding post-op fusion logic (which means no negative perf impact for users).

Reply:
Removed completely, because it is replaced by the suggested text.